c10/core/Event.h

```diff
@@ -0,0 +1,145 @@
+#pragma once
+
+#include "ATen/Context.h"
+#include "c10/util/Optional.h"
+#include "c10/util/Exception.h"
+#include "c10/core/Stream.h"
+
+#include <memory>
+#include <limits>
+#include <atomic>
+#include <utility>
+
+namespace at { namespace cuda {  namespace detail {
+    struct CUDAHooks;
+}}} // at::cuda::detail
+
+namespace c10 {
+
+/*
+* A generic Event capable of representing device-specific events when the appropriate
+* backend is loaded. It is designed to dispatch calls to different backends
+* based on the Stream the event is recorded on.
+*
+* This class is not thread-safe and is movable but not copyable. 
+*
+*/
+struct Event final {
+
+static constexpr unsigned int DEFAULT_FLAGS = std::numeric_limits<unsigned int>::max();
+friend class at::cuda::detail::CUDAHooks;
+
+// Constructor optionally accepts flags.
+// Note: DEFAULT_FLAGS reflects PyTorch's defaults, not the default values
+// for constructing a particular backend's events. For example, CUDA events
+// are constructed by PyTorch with the cudaEventDisableTiming flag, which is
+// not the default constructor behavior for a CUDA event. To obtain a CUDA event
+// with the CUDA default flags pass zero (=cudaEventDefault) explicitly. 
+Event(const unsigned int flags = DEFAULT_FLAGS) 
```

Comment:
Don't you need to know what the device of the Event is at construction time? We don't have to actually make this work for multiple backends to start, but we should at least make sure the API is right (and not hard code CUDA).

Reply:
We could allow (or force) the user to specify the backend when the Event is constructed for "type checking," but it's not strictly necessary. Per Zach's point below we could abstract event flags in an enum to make them backend-agnostic and then lazily construct events on the appropriate backend with the appropriate flags. I think generalizing the flags and optionally allowing the backend to be specified will be the best of both worlds. Sound good?